spoc/query_engine.py

```diff
@@ -0,0 +1,254 @@
+        return self.__repr__()
+
+
+class Snipper:
+    """
+    This class represents a snipper used for contact selection.
+    It provides methods to validate the filter against a data schema,
+    convert data to a duckdb relation, construct a filter string,
+    and apply the filter to the data.
+
+    Attributes:
+        _regions (List[pd.DataFrame]): The regions to be used for filtering.
+        _anchor_mode (Anchor): The anchor mode for filtering.
+    """
+
+    def __init__(self, regions: List[pd.DataFrame], anchor_mode: Anchor) -> None:
+        # add ids to regions if they don't exist
+        if "id" not in regions.columns:
+            regions["id"] = range(len(regions))
+        self._regions = RegionSchema.validate(regions)
+        self._anchor_mode = anchor_mode
+
+    def validate(self, data_schema: GenomicDataSchema) -> None:
+        """Validate the filter against the data schema"""
+        # check whether an anchor is specified that is not in the data
+        if self._anchor_mode.anchors is not None:
+            if not all(
+                anchor in data_schema.get_position_fields().keys()
+                for anchor in self._anchor_mode.anchors
+            ):
+                raise ValueError(
+                    "An anchor is specified that is not in the data schema."
+                )
+
+    def _convert_to_duckdb(
+        self,
+        data: Union[pd.DataFrame, dd.DataFrame, duckdb.DuckDBPyRelation],
+    ) -> duckdb.DuckDBPyRelation:
+        """
+        Converts the data to a duckdb relation.
+
+        Parameters:
+            data (Union[pd.DataFrame, dd.DataFrame, duckdb.DuckDBPyRelation]): The input data to be converted.
+
+        Returns:
+            duckdb.DuckDBPyRelation: The converted duckdb relation.
+        """
+        if isinstance(data, dd.DataFrame):
+            data = data.compute()
+        return duckdb.from_df(data, connection=DUCKDB_CONNECTION)
+
+    def _contstruct_filter(self, position_fields: Dict[int, List[str]]) -> str:
+        """Constructs the filter string.
+
+        Args:
+            position_fields (List[str]): List of position fields.
+
+        Returns:
+            str: The constructed filter string.
+
+        Raises:
+            NotImplementedError: If the length of fields is not equal to 3.
+        """
+        query_strings = []
+        join_string = " or " if self._anchor_mode.mode == "ANY" else " and "
+        # subset on anchor regions
+        if self._anchor_mode.anchors is not None:
+            subset_positions = [
+                position_fields[anchor] for anchor in self._anchor_mode.anchors
+            ]
+        else:
+            subset_positions = list(position_fields.values())
+        for fields in subset_positions:
+            # check two or three-way definition of fields
+            if len(fields) == 3:
+                chrom, start, end = fields
+                output_string = f"""(data.{chrom} = regions.chrom and
```

Comment:
Yes, it does not. It was somewhat by design to keep the interface simple. Do you expect a lot of use cases when regions are smaller than bins? Keep in mind that the regions here are intended to act like the windows that we defined for the snipping in cooltools to provide a view onto the genomes.

Reply:
I can imagine smth like this, e.g. find all multiway contacts that have one fragment overlapping a specific promoter (or a set of promoters, or CTCF sites). For our cases it might be not a frequent use case, but for general audience -- yeah. So I suggest we keep it stringent. Anyway, it's just adding another subclause `or regions.start between data.start and data.end`, shouldn't be too complicated.

I am curious, though, how does duckdb evaluate OR clauses? In python, if A from (A or B) is true, B is not evaluated. If duckdb does the same, that would be cool. So if we expect many overlaps, than most of the times in this case duckdb will evaluate only first two subclauses, so we shouldn't drop in performance much. But it would be interesting to test (monkey-patching?).

Reply:
Cool, so now we have only this comment left (and maybe arrows in the UML diagram, but that's not very important).

Reply:
Alright! I added the additional condition :) I am almost certain that duckdb does lazy evaluation of OR clauses, so I don't think there is a performance penalty!

Reply:
Yay, thanks!